File: delivery/php-sdk/client/ApiResponse.php
<?php

namespace Maiyatian\Delivery\PhpSdk\Client;

use Maiyatian\Delivery\PhpSdk\Consts\Consts;

/**
 * 类型安全的API响应包装器
 * 对应 Go SDK 中的 ApiResponse[T any] 泛型
 * PHP中使用模板注释和运行时类型检查实现类似功能
 * 
 * @template T
 * @author SDK Generator  
 * @version 1.0.0
 * @since 1.0.0
 */
class ApiResponse
{
    /**
     * 业务状态码：200 表示成功，其他表示失败
     * @var int
     */
    private $code;
    
    /**
     * 返回消息：成功时为 "ok"，失败时为错误描述
     * @var string
     */
    private $message;
    
    /**
     * 业务数据，已解析为具体的PHP对象或数组
     * @var T|null
     */
    private $data;
    
    /**
     * 原始响应对象
     * @var HTTPResponse
     */
    private $rawResponse;
    
    /**
     * 响应数据的类型信息
     * @var string|null
     */
    private $dataType;
    
    // ==================== 构造函数 ====================
    
    /**
     * ApiResponse 构造函数
     * 
     * @param int $code 状态码
     * @param string $message 响应消息
     * @param T|null $data 业务数据
     * @param HTTPResponse|null $rawResponse 原始响应
     */
    public function __construct(int $code, string $message, $data = null, ?HTTPResponse $rawResponse = null)
    {
        $this->code = $code;
        $this->message = $message;
        $this->data = $data;
        $this->rawResponse = $rawResponse;
        $this->dataType = is_object($data) ? get_class($data) : gettype($data);
    }
    
    // ==================== 静态工厂方法 ====================
    
    /**
     * 创建成功响应
     * 对应 Go SDK 的成功响应创建
     * 
     * @template U
     * @param U $data 响应数据
     * @param string $message 成功消息
     * @return ApiResponse<U>
     */
    public static function success($data, string $message = 'ok'): self
    {
        return new self(Consts::SUCCESS_CODE, $message, $data);
    }
    
    /**
     * 创建错误响应
     * 对应 Go SDK 的错误响应创建
     * 
     * @param int $code 错误码
     * @param string $message 错误消息
     * @return ApiResponse<null>
     */
    public static function error(int $code, string $message): self
    {
        return new self($code, $message, null);
    }
    
    /**
     * 从HTTPResponse转换创建ApiResponse
     * 对应 Go SDK 中 RequestWithApiClient 的转换逻辑
     * 
     * @template U
     * @param HTTPResponse $response HTTP响应对象
     * @param string|callable|null $dataClass 数据类型或转换函数
     * @return ApiResponse<U>
     * @throws \InvalidArgumentException
     */
    public static function from(HTTPResponse $response, $dataClass = null): self
    {
        $data = null;
        
        // 解析响应数据
        if (!empty($response->Data)) {
            try {
                if ($dataClass === null) {
                    // 不指定类型，直接解析为数组
                    $data = json_decode($response->Data, true);
                    if (json_last_error() !== JSON_ERROR_NONE) {
                        throw new \InvalidArgumentException('JSON解析失败: ' . json_last_error_msg());
                    }
                } elseif (is_string($dataClass)) {
                    // 指定具体类，尝试实例化
                    $arrayData = json_decode($response->Data, true);
                    if (json_last_error() !== JSON_ERROR_NONE) {
                        throw new \InvalidArgumentException('JSON解析失败: ' . json_last_error_msg());
                    }
                    
                    if (class_exists($dataClass)) {
                        $data = self::instantiateFromArray($dataClass, $arrayData);
                    } else {
                        $data = $arrayData;
                    }
                } elseif (is_callable($dataClass)) {
                    // 使用自定义转换函数
                    $arrayData = json_decode($response->Data, true);
                    if (json_last_error() !== JSON_ERROR_NONE) {
                        throw new \InvalidArgumentException('JSON解析失败: ' . json_last_error_msg());
                    }
                    $data = $dataClass($arrayData);
                } else {
                    throw new \InvalidArgumentException('无效的数据类型参数');
                }
            } catch (\Exception $e) {
                throw new \InvalidArgumentException('数据转换失败: ' . $e->getMessage(), 0, $e);
            }
        }
        
        return new self($response->Code, $response->Message, $data, $response);
    }
    
    // ==================== Getter 方法 ====================
    
    /**
     * 获取状态码
     * 
     * @return int
     */
    public function getCode(): int
    {
        return $this->code;
    }
    
    /**
     * 获取响应消息
     * 
     * @return string
     */
    public function getMessage(): string
    {
        return $this->message;
    }
    
    /**
     * 获取响应数据
     * 
     * @return T|null
     */
    public function getData()
    {
        return $this->data;
    }
    
    /**
     * 获取原始响应对象
     * 
     * @return HTTPResponse|null
     */
    public function getRawResponse(): ?HTTPResponse
    {
        return $this->rawResponse;
    }
    
    /**
     * 获取数据类型信息
     * 
     * @return string|null
     */
    public function getDataType(): ?string
    {
        return $this->dataType;
    }
    
    // ==================== 状态检查方法 ====================
    
    /**
     * 检查是否是成功响应
     * 对应 Go SDK 的成功状态检查
     * 
     * @return bool
     */
    public function isSuccess(): bool
    {
        return $this->code === Consts::SUCCESS_CODE;
    }
    
    /**
     * 检查是否是错误响应
     * 
     * @return bool
     */
    public function isError(): bool
    {
        return !$this->isSuccess();
    }
    
    /**
     * 检查是否有数据
     * 
     * @return bool
     */
    public function hasData(): bool
    {
        return $this->isSuccess() && $this->data !== null;
    }
    
    // ==================== 函数式编程方法（类似Go泛型） ====================
    
    /**
     * 转换数据类型（map操作）
     * 对应 Go SDK 泛型的类型转换功能
     * 
     * @template U
     * @param callable $mapper 转换函数
     * @return ApiResponse<U>
     */
    public function map(callable $mapper): self
    {
        if ($this->isError() || $this->data === null) {
            return new self($this->code, $this->message, null, $this->rawResponse);
        }
        
        try {
            $mappedData = $mapper($this->data);
            return new self($this->code, $this->message, $mappedData, $this->rawResponse);
        } catch (\Exception $e) {
            return self::error(500, '数据转换失败: ' . $e->getMessage());
        }
    }
    
    /**
     * 过滤响应（filter操作）
     * 
     * @param callable $predicate 过滤条件
     * @return ApiResponse<T>
     */
    public function filter(callable $predicate): self
    {
        if ($this->isError()) {
            return $this;
        }
        
        try {
            if ($predicate($this)) {
                return $this;
            } else {
                return self::error(400, '响应不满足过滤条件');
            }
        } catch (\Exception $e) {
            return self::error(500, '过滤条件检查失败: ' . $e->getMessage());
        }
    }
    
    /**
     * 获取数据或返回默认值
     * 对应 Go SDK 的安全数据获取
     * 
     * @param T $defaultValue 默认值
     * @return T
     */
    public function getDataOrDefault($defaultValue)
    {
        if ($this->isSuccess() && $this->data !== null) {
            return $this->data;
        }
        return $defaultValue;
    }
    
    /**
     * 获取数据或抛出异常
     * 
     * @return T
     * @throws ApiException
     */
    public function getDataOrThrow()
    {
        if ($this->isSuccess() && $this->data !== null) {
            return $this->data;
        }
        
        throw new ApiException(
            sprintf('API错误 [%d]: %s', $this->code, $this->message),
            $this->code
        );
    }
    
    // ==================== 配送业务特定方法 ====================
    
    /**
     * 检查是否是配送业务成功响应
     * 
     * @return bool
     */
    public function isDeliverySuccess(): bool
    {
        return $this->isSuccess() && $this->data !== null;
    }
    
    /**
     * 获取配送状态
     * 专门用于配送业务的状态获取
     * 
     * @return int|null
     */
    public function getDeliveryStatus(): ?int
    {
        if ($this->isError() || !is_array($this->data)) {
            return null;
        }
        
        return $this->data['status'] ?? $this->data['delivery_status'] ?? null;
    }
    
    /**
     * 检查是否是位置变更响应
     * 
     * @return bool
     */
    public function isLocationChange(): bool
    {
        if ($this->isError() || !is_array($this->data)) {
            return false;
        }
        
        return isset($this->data['latitude']) && isset($this->data['longitude']);
    }
    
    // ==================== 私有辅助方法 ====================
    
    /**
     * 从数组实例化对象
     * 
     * @param string $className 类名
     * @param array $data 数据数组
     * @return object
     */
    private static function instantiateFromArray(string $className, array $data)
    {
        $reflection = new \ReflectionClass($className);
        $instance = $reflection->newInstanceWithoutConstructor();
        
        foreach ($data as $key => $value) {
            $property = null;
            
            // 尝试直接访问属性
            if ($reflection->hasProperty($key)) {
                $property = $reflection->getProperty($key);
                if (!$property->isPublic()) {
                    $property->setAccessible(true);
                }
                $property->setValue($instance, $value);
            } elseif (method_exists($instance, 'set' . ucfirst($key))) {
                // 尝试setter方法
                $instance->{'set' . ucfirst($key)}($value);
            } elseif (property_exists($instance, $key)) {
                // 动态属性
                $instance->$key = $value;
            }
        }
        
        return $instance;
    }
    
    // ==================== 魔术方法 ====================
    
    /**
     * 字符串表示
     * 
     * @return string
     */
    public function __toString(): string
    {
        return sprintf(
            'ApiResponse{code=%d, message="%s", hasData=%s, dataType=%s}',
            $this->code,
            $this->message,
            $this->hasData() ? 'true' : 'false',
            $this->dataType ?? 'null'
        );
    }
    
    /**
     * 序列化为数组
     * 
     * @return array
     */
    public function toArray(): array
    {
        return [
            'code' => $this->code,
            'message' => $this->message,
            'data' => $this->data,
            'hasData' => $this->hasData(),
            'isSuccess' => $this->isSuccess(),
            'isDeliverySuccess' => $this->isDeliverySuccess(),
            'deliveryStatus' => $this->getDeliveryStatus(),
            'isLocationChange' => $this->isLocationChange()
        ];
    }
}

/**
 * API异常类
 */
class ApiException extends \Exception
{
    public function __construct(string $message, int $code = 0, ?\Throwable $previous = null)
    {
        parent::__construct($message, $code, $previous);
    }
}

File: delivery/java-sdk/src/main/java/com/maiyatian/delivery/sdk/models/types/ApiResponse.java
/*
 * 麦芽田配送开放平台 Java SDK
 * Copyright (c) 2025 麦芽田开放平台
 * MIT License
 */

package com.maiyatian.delivery.sdk.models.types;

import com.maiyatian.delivery.sdk.consts.Constants;
import java.util.Objects;

/**
 * 类型安全的API响应包装器
 * <p>
 * 对应Go SDK中的ApiResponse[T any]结构，提供类型安全的响应数据访问
 * 实现从Response到强类型数据的自动转换
 * </p>
 * 
 * <h3>功能特性：</h3>
 * <ul>
 * <li><strong>类型安全</strong>：编译时类型检查，避免类型转换错误</li>
 * <li><strong>空值处理</strong>：优雅处理null和空数据的情况</li>
 * <li><strong>异常处理</strong>：提供getDataOrThrow()方法处理异常场景</li>
 * <li><strong>链式操作</strong>：支持流畅的API调用模式</li>
 * </ul>
 * 
 * <h3>对应关系：</h3>
 * <ul>
 * <li>Go: {@code type ApiResponse[T any] struct { Code int; Message string; Data *T }} → 
 *     Java: {@code class ApiResponse<T> { int code; String message; T data }}</li>
 * <li>Go: {@code response.Data} → Java: {@code response.getData()}</li>
 * </ul>
 * 
 * <h3>使用示例：</h3>
 * <pre>{@code
 * // 成功响应处理
 * ApiResponse<AccessTokenResp> response = deliverySender.accessToken(token, request);
 * if (response.isSuccess()) {
 *     AccessTokenResp tokenData = response.getDataOrThrow();
 *     System.out.println("访问令牌: " + tokenData.getToken());
 * }
 * 
 * // 错误响应处理
 * ApiResponse<String> response = deliverySender.someApi(data);
 * if (!response.isSuccess()) {
 *     System.err.println("API调用失败: " + response.getMessage());
 * }
 * 
 * // 链式操作
 * String token = deliverySender.accessToken(authData)
 *     .filter(ApiResponse::isSuccess)
 *     .map(ApiResponse::getDataOrThrow)
 *     .map(AccessTokenResp::getToken)
 *     .orElseThrow(() -> new RuntimeException("获取令牌失败"));
 * }</pre>
 * 
 * @param <T> 响应数据类型
 * @author SDK Generator
 * @version 1.0.0
 * @since 1.0.0
 */
public class ApiResponse<T> {
    
    /**
     * 业务状态码
     * <p>200表示成功，其他值表示失败</p>
     */
    private int code;
    
    /**
     * 响应消息
     * <p>成功时为"ok"，失败时为错误描述</p>
     */
    private String message;
    
    /**
     * 业务数据
     * <p>类型安全的强类型数据，可能为null</p>
     */
    private T data;
    
    // ==================== 构造函数 ====================
    
    /**
     * 默认构造函数
     */
    public ApiResponse() {
    }
    
    /**
     * 完整构造函数
     * 
     * @param code 业务状态码
     * @param message 响应消息
     * @param data 业务数据
     */
    public ApiResponse(int code, String message, T data) {
        this.code = code;
        this.message = message;
        this.data = data;
    }
    
    // ==================== 静态工厂方法 ====================
    
    /**
     * 创建成功响应
     * <p>
     * 工厂方法，创建一个表示成功的ApiResponse实例
     * </p>
     * 
     * @param <T> 数据类型
     * @param data 响应数据
     * @return 成功的ApiResponse实例
     */
    public static <T> ApiResponse<T> success(T data) {
        return new ApiResponse<>(Constants.SUCCESS_CODE, "ok", data);
    }
    
    /**
     * 创建成功响应（无数据）
     * <p>
     * 用于只返回成功状态，不返回具体数据的API
     * </p>
     * 
     * @return 成功的空数据ApiResponse实例
     */
    public static ApiResponse<Void> success() {
        return new ApiResponse<>(Constants.SUCCESS_CODE, "ok", null);
    }
    
    /**
     * 创建错误响应
     * <p>
     * 工厂方法，创建一个表示失败的ApiResponse实例
     * </p>
     * 
     * @param <T> 数据类型
     * @param code 错误状态码
     * @param message 错误消息
     * @return 失败的ApiResponse实例
     */
    public static <T> ApiResponse<T> error(int code, String message) {
        return new ApiResponse<>(code, message, null);
    }
    
    /**
     * 从Response转换为ApiResponse
     * <p>
     * 对应Go SDK中的FormatData()方法功能
     * 将原始Response转换为类型安全的ApiResponse
     * </p>
     * 
     * @param <T> 目标数据类型
     * @param response 原始响应
     * @param dataClass 目标数据类型的Class对象
     * @return 转换后的ApiResponse
     * @throws RuntimeException 如果数据解析失败
     */
    public static <T> ApiResponse<T> from(Response response, Class<T> dataClass) {
        ApiResponse<T> apiResponse = new ApiResponse<>();
        apiResponse.setCode(response.getCode());
        apiResponse.setMessage(response.getMessage());
        
        // 解析数据到指定类型
        if (response.getData() != null && !response.getData().trim().isEmpty()) {
            if (dataClass != Void.class && dataClass != void.class) {
                try {
                    T parsedData = response.parseData(dataClass);
                    apiResponse.setData(parsedData);
                } catch (Exception e) {
                    throw new RuntimeException("Failed to parse response data to " + dataClass.getSimpleName(), e);
                }
            }
        }
        
        return apiResponse;
    }
    
    /**
     * 创建空数据响应
     * <p>
     * 专门用于处理无返回数据的API响应
     * </p>
     * 
     * @param response 原始响应
     * @return 空数据的ApiResponse
     */
    @SuppressWarnings("unchecked")
    public static ApiResponse<Void> fromEmpty(Response response) {
        return (ApiResponse<Void>) from(response, Void.class);
    }
    
    // ==================== 业务方法 ====================
    
    /**
     * 检查响应是否成功
     * <p>
     * 根据状态码判断API调用是否成功
     * </p>
     * 
     * @return true表示成功，false表示失败
     */
    public boolean isSuccess() {
        return code == Constants.SUCCESS_CODE;
    }
    
    /**
     * 检查响应是否失败
     * <p>
     * 与isSuccess()相反的逻辑
     * </p>
     * 
     * @return true表示失败，false表示成功
     */
    public boolean isError() {
        return !isSuccess();
    }
    
    /**
     * 获取数据，失败时抛出异常
     * <p>
     * 安全的数据获取方法，确保只在成功时返回数据
     * 失败时抛出包含错误信息的RuntimeException
     * </p>
     * 
     * <h3>使用场景：</h3>
     * <ul>
     * <li>确保API调用成功的场景</li>
     * <li>需要快速失败的业务逻辑</li>
     * <li>链式调用中的中间环节</li>
     * </ul>
     * 
     * @return 响应数据
     * @throws RuntimeException 如果响应失败或数据为null
     */
    public T getDataOrThrow() {
        if (isError()) {
            throw new RuntimeException("API调用失败 [" + code + "]: " + message);
        }
        if (data == null) {
            throw new RuntimeException("响应数据为空");
        }
        return data;
    }
    
    /**
     * 获取数据，失败时返回默认值
     * <p>
     * 安全的数据获取方法，失败时返回指定的默认值
     * 适用于可以容忍失败的业务场景
     * </p>
     * 
     * @param defaultValue 默认值
     * @return 响应数据或默认值
     */
    public T getDataOrDefault(T defaultValue) {
        if (isSuccess() && data != null) {
            return data;
        }
        return defaultValue;
    }
    
    /**
     * 检查是否有数据
     * <p>
     * 同时检查响应成功和数据不为null
     * </p>
     * 
     * @return true表示有有效数据，false表示无数据
     */
    public boolean hasData() {
        return isSuccess() && data != null;
    }
    
    /**
     * 转换数据类型
     * <p>
     * 提供类型转换功能，支持链式操作
     * </p>
     * 
     * @param <R> 目标类型
     * @param mapper 转换函数
     * @return 转换后的ApiResponse
     */
    public <R> ApiResponse<R> map(java.util.function.Function<T, R> mapper) {
        if (isError() || data == null) {
            return new ApiResponse<>(code, message, null);
        }
        
        try {
            R mappedData = mapper.apply(data);
            return new ApiResponse<>(code, message, mappedData);
        } catch (Exception e) {
            return ApiResponse.error(500, "数据转换失败: " + e.getMessage());
        }
    }
    
    /**
     * 过滤响应
     * <p>
     * 支持基于条件的响应过滤
     * </p>
     * 
     * @param predicate 过滤条件
     * @return 过滤后的ApiResponse
     */
    public ApiResponse<T> filter(java.util.function.Predicate<ApiResponse<T>> predicate) {
        if (predicate.test(this)) {
            return this;
        }
        return ApiResponse.error(400, "响应不满足过滤条件");
    }
    
    // ==================== Getter/Setter ====================
    
    /**
     * 获取业务状态码
     * 
     * @return 状态码
     */
    public int getCode() {
        return code;
    }
    
    /**
     * 设置业务状态码
     * 
     * @param code 状态码
     */
    public void setCode(int code) {
        this.code = code;
    }
    
    /**
     * 获取响应消息
     * 
     * @return 消息内容
     */
    public String getMessage() {
        return message;
    }
    
    /**
     * 设置响应消息
     * 
     * @param message 消息内容
     */
    public void setMessage(String message) {
        this.message = message;
    }
    
    /**
     * 获取业务数据
     * 
     * @return 业务数据（可能为null）
     */
    public T getData() {
        return data;
    }
    
    /**
     * 设置业务数据
     * 
     * @param data 业务数据
     */
    public void setData(T data) {
        this.data = data;
    }
    
    // ==================== Object方法 ====================
    
    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        ApiResponse<?> that = (ApiResponse<?>) o;
        return code == that.code &&
               Objects.equals(message, that.message) &&
               Objects.equals(data, that.data);
    }
    
    @Override
    public int hashCode() {
        return Objects.hash(code, message, data);
    }
    
    @Override
    public String toString() {
        return "ApiResponse{" +
                "code=" + code +
                ", message='" + message + '\'' +
                ", data=" + data +
                ", hasData=" + hasData() +
                '}';
    }
}